File: src/Butil/Bit.Butil/Publics/Url.cs
using System;
using System.Diagnostics.CodeAnalysis;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps the <see href="https://developer.mozilla.org/en-US/docs/Web/API/URL">URL</see>,
/// <see href="https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams">URLSearchParams</see>
/// and <see href="https://developer.mozilla.org/en-US/docs/Web/API/URLPattern">URLPattern</see> APIs.
/// </summary>
/// <remarks>
/// The first two overlap <see cref="System.Uri"/>, and where they do, <c>System.Uri</c> is the better
/// tool - it needs no interop call. Two reasons to come here anyway:
/// <list type="bullet">
/// <item>The browser follows the WHATWG URL standard, and <c>System.Uri</c> follows RFC 3986. They
/// disagree in ordinary cases - on trailing dots in a host, on how far <c>..</c> may climb, on which
/// characters get percent-encoded - so when a URL has to match what the browser will do with it
/// (a fetch, a redirect, an origin comparison), asking the browser is the only way to be sure.</item>
/// <item><see cref="TestPattern"/> and <see cref="MatchPattern"/> have no .NET equivalent at all.
/// <c>URLPattern</c> is a route matcher with named parameters, wildcards and per-component patterns -
/// <c>/books/:id</c> against a URL - which is a thing apps write by hand over and over.</item>
/// </list>
/// </remarks>
[ButilService(typeof(Url))]
public class Url(IJSRuntime js)
{
    /// <summary>True when the runtime exposes <c>URL</c> - which is every runtime that runs Blazor.</summary>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.url.isSupported");

    /// <summary>
    /// Whether the browser can parse this URL - <c>URL.canParse()</c>, without the exception the
    /// constructor would throw.
    /// </summary>
    /// <param name="url">An absolute URL, or a reference to resolve against <paramref name="baseUrl"/>.</param>
    /// <param name="baseUrl">Optional base for a relative reference.</param>
    public ValueTask<bool> CanParse(string url, string? baseUrl = null)
        => js.Invoke<bool>("BitButil.url.canParse", url, baseUrl);

    /// <summary>
    /// Splits a URL into its components exactly as the browser does.
    /// </summary>
    /// <param name="url">An absolute URL, or a reference to resolve against <paramref name="baseUrl"/>.</param>
    /// <param name="baseUrl">Optional base for a relative reference.</param>
    /// <returns>The parts, or null when the URL doesn't parse.</returns>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(UrlParts))]
    public ValueTask<UrlParts?> Parse(string url, string? baseUrl = null)
        => js.Invoke<UrlParts?>("BitButil.url.parse", url, baseUrl);

    /// <summary>
    /// Resolves a possibly relative reference against a base, the way the browser's own resolver
    /// does it (including the scheme-relative <c>//host/path</c> and over-climbing <c>..</c> cases).
    /// </summary>
    /// <returns>The absolute URL, or null when the result doesn't parse.</returns>
    public ValueTask<string?> Resolve(string url, string baseUrl)
        => js.Invoke<string?>("BitButil.url.resolve", url, baseUrl);

    /// <summary>
    /// Parses a query string into its parameters.
    /// </summary>
    /// <param name="query">A query string, with or without the leading <c>?</c>.</param>
    /// <returns>
    /// One entry per parameter, in order, keeping repeats: <c>?tag=a&amp;tag=b</c> is two entries,
    /// which is why this is a list rather than a dictionary.
    /// </returns>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(UrlQueryParameter))]
    public ValueTask<UrlQueryParameter[]> ParseQuery(string query)
        => js.Invoke<UrlQueryParameter[]>("BitButil.url.parseQuery", query);

    /// <summary>
    /// Builds a query string from parameters, percent-encoding each one the way the browser does.
    /// </summary>
    /// <returns>The query string without a leading <c>?</c>.</returns>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(UrlQueryParameter))]
    public ValueTask<string> BuildQuery(UrlQueryParameter[] parameters)
    {
        ArgumentNullException.ThrowIfNull(parameters);
        // new object?[] { parameters } wraps the array as a single JS argument; passing it bare would
        // spread it, since the params array is object?[] and an array of a reference type converts to it.
        return js.Invoke<string>("BitButil.url.buildQuery", new object?[] { parameters });
    }

    /// <summary>
    /// Every value of one repeated parameter, in order. Empty when the parameter isn't there.
    /// </summary>
    public ValueTask<string[]> GetQueryValues(string query, string key)
        => js.Invoke<string[]>("BitButil.url.getQueryValues", query, key);

    /// <summary>Replaces a URL's whole query string. Null when the URL doesn't parse.</summary>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(UrlQueryParameter))]
    public ValueTask<string?> SetQuery(string url, UrlQueryParameter[] parameters)
    {
        ArgumentNullException.ThrowIfNull(parameters);
        return js.Invoke<string?>("BitButil.url.setQuery", url, parameters);
    }

    /// <summary>
    /// Appends parameters to a URL's query, keeping any that are already there - including one of
    /// the same name.
    /// </summary>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(UrlQueryParameter))]
    public ValueTask<string?> AppendQuery(string url, UrlQueryParameter[] parameters)
    {
        ArgumentNullException.ThrowIfNull(parameters);
        return js.Invoke<string?>("BitButil.url.appendQuery", url, parameters);
    }

    /// <summary>Removes every parameter with any of these names from a URL's query.</summary>
    public ValueTask<string?> RemoveQuery(string url, params string[] keys)
        => js.Invoke<string?>("BitButil.url.removeQuery", url, keys ?? []);

    /// <summary>
    /// Sorts a URL's query parameters by name. Useful for a cache key or a signature, where two URLs
    /// that differ only in parameter order should come out identical.
    /// </summary>
    public ValueTask<string?> SortQuery(string url) => js.Invoke<string?>("BitButil.url.sortQuery", url);

    /// <summary>True when the runtime exposes <c>URLPattern</c>.</summary>
    /// <remarks>
    /// Newer than the rest of this class: Chromium and Safari have it, Firefox shipped it later.
    /// During prerender/SSR this returns false rather than throwing - see <see cref="IsSupported"/>.
    /// </remarks>
    public ValueTask<bool> IsPatternSupported() => js.Invoke<bool>("BitButil.url.isPatternSupported");

    /// <summary>
    /// Whether a pattern itself compiles. A malformed pattern and a URL that simply doesn't match
    /// both look like "no match" otherwise.
    /// </summary>
    public ValueTask<bool> IsPatternValid(string pattern, string? baseUrl = null)
        => js.Invoke<bool>("BitButil.url.isPatternValid", pattern, baseUrl);

    /// <summary>
    /// Whether a URL matches a pattern.
    /// </summary>
    /// <param name="pattern">
    /// A URL pattern: <c>"/books/:id"</c>, <c>"https://*.example.com/*"</c>,
    /// <c>"/files/:name.:ext"</c>. Named parameters start with <c>:</c>, <c>*</c> is a wildcard, and
    /// a group in <c>{}</c> can be made optional with <c>?</c>.
    /// </param>
    /// <param name="url">The URL to test.</param>
    /// <param name="baseUrl">Base the pattern is relative to - required for a path-only pattern like <c>"/books/:id"</c>.</param>
    /// <returns>False when the pattern is invalid or <c>URLPattern</c> isn't supported, as well as when it simply doesn't match.</returns>
    public ValueTask<bool> TestPattern(string pattern, string url, string? baseUrl = null)
        => js.Invoke<bool>("BitButil.url.patternTest", pattern, baseUrl, url);

    /// <summary>
    /// Matches a URL against a pattern and returns what the named parameters captured - the routing
    /// case: <c>"/books/:id"</c> against <c>"/books/42"</c> gives <c>id = "42"</c>.
    /// </summary>
    /// <param name="pattern">The pattern - see <see cref="TestPattern"/>.</param>
    /// <param name="url">The URL to match.</param>
    /// <param name="baseUrl">Base the pattern is relative to - required for a path-only pattern.</param>
    /// <returns>The match, or null when the URL doesn't match, the pattern is invalid, or <c>URLPattern</c> isn't supported.</returns>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(UrlPatternMatch))]
    public ValueTask<UrlPatternMatch?> MatchPattern(string pattern, string url, string? baseUrl = null)
        => js.Invoke<UrlPatternMatch?>("BitButil.url.patternExec", pattern, baseUrl, url);
}

File: src/Butil/Bit.Butil/Scripts/url.ts
var BitButil = (window as any).BitButil = (window as any).BitButil || {};

(function (butil: any) {
    function parseUrl(url: string, base?: string | null) {
        try { return base ? new URL(url, base) : new URL(url); } catch { return null; }
    }

    function toParts(parsed: URL) {
        return {
            href: parsed.href,
            origin: parsed.origin,
            protocol: parsed.protocol,
            username: parsed.username,
            password: parsed.password,
            host: parsed.host,
            hostname: parsed.hostname,
            port: parsed.port,
            pathname: parsed.pathname,
            search: parsed.search,
            hash: parsed.hash
        };
    }

    // URLPattern takes either a pattern string or an object of per-component patterns; the string
    // form needs a base to resolve against unless it is absolute, which is why both are accepted.
    function makePattern(pattern: string, base?: string | null) {
        const ctor = (window as any).URLPattern;
        if (typeof ctor !== 'function') return null;
        try { return base ? new ctor(pattern, base) : new ctor(pattern); } catch { return null; }
    }

    // The named groups of every component in one map. A pattern names each group once across the
    // whole URL, so flattening loses nothing and spares the caller a component-shaped object; the
    // component a group came from is still in `components` below.
    function flattenGroups(result: any) {
        const groups: { [key: string]: string | null } = {};
        for (const component of ['protocol', 'username', 'password', 'hostname', 'port', 'pathname', 'search', 'hash']) {
            const matched = result[component];
            if (!matched?.groups) continue;
            for (const key of Object.keys(matched.groups)) {
                // Positional groups are keyed '0', '1', ... in every component, so a later component
                // would overwrite an earlier one's. First writer wins, matching left-to-right order.
                if (key in groups) continue;
                groups[key] = matched.groups[key] ?? null;
            }
        }
        return groups;
    }

    butil.url = {
        isSupported() { return typeof URL === 'function'; },
        canParse(url: string, base?: string | null) {
            // URL.canParse is the cheap path; older runtimes only have the throwing constructor.
            const can = (URL as any).canParse;
            if (typeof can === 'function') { try { return base ? can(url, base) : can(url); } catch { return false; } }
            return parseUrl(url, base) !== null;
        },
        parse(url: string, base?: string | null) {
            const parsed = parseUrl(url, base);
            return parsed ? toParts(parsed) : null;
        },
        // Absolute form of a possibly relative reference, the way the browser's own resolver does it
        // (including the '..' and scheme-relative cases System.Uri handles differently).
        resolve(url: string, base: string) {
            const parsed = parseUrl(url, base);
            return parsed ? parsed.href : null;
        },

        parseQuery(query: string) {
            const params = new URLSearchParams(query ?? '');
            // As a flat list rather than a map: a query string may repeat a key, and a map would
            // silently drop all but one of them.
            const pairs: { key: string, value: string }[] = [];
            params.forEach((value, key) => pairs.push({ key, value }));
            return pairs;
        },
        buildQuery(pairs: any[]) {
            const params = new URLSearchParams();
            for (const pair of pairs ?? []) params.append(pair.key, pair.value ?? '');
            return params.toString();
        },
        getQueryValues(query: string, key: string) {
            return new URLSearchParams(query ?? '').getAll(key);
        },
        // Whole-URL edits, so the caller never has to re-assemble the parts by hand.
        setQuery(url: string, pairs: any[]) {
            const parsed = parseUrl(url);
            if (!parsed) return null;
            parsed.search = butil.url.buildQuery(pairs);
            return parsed.href;
        },
        appendQuery(url: string, pairs: any[]) {
            const parsed = parseUrl(url);
            if (!parsed) return null;
            for (const pair of pairs ?? []) parsed.searchParams.append(pair.key, pair.value ?? '');
            return parsed.href;
        },
        removeQuery(url: string, keys: string[]) {
            const parsed = parseUrl(url);
            if (!parsed) return null;
            for (const key of keys ?? []) parsed.searchParams.delete(key);
            return parsed.href;
        },
        sortQuery(url: string) {
            const parsed = parseUrl(url);
            if (!parsed) return null;
            parsed.searchParams.sort();
            return parsed.href;
        },

        isPatternSupported() { return typeof (window as any).URLPattern === 'function'; },
        patternTest(pattern: string, base: string | null, url: string) {
            const compiled = makePattern(pattern, base);
            if (!compiled) return false;
            try { return compiled.test(url); } catch { return false; }
        },
        patternExec(pattern: string, base: string | null, url: string) {
            const compiled = makePattern(pattern, base);
            if (!compiled) return null;
            let result: any;
            try { result = compiled.exec(url); } catch { return null; }
            if (!result) return null;
            return {
                protocol: result.protocol?.input ?? '',
                username: result.username?.input ?? '',
                password: result.password?.input ?? '',
                hostname: result.hostname?.input ?? '',
                port: result.port?.input ?? '',
                pathname: result.pathname?.input ?? '',
                search: result.search?.input ?? '',
                hash: result.hash?.input ?? '',
                groups: flattenGroups(result)
            };
        },
        // Whether the pattern itself compiles - a bad pattern and a URL that simply doesn't match
        // both come back as "no match" otherwise.
        isPatternValid(pattern: string, base: string | null) {
            return makePattern(pattern, base) !== null;
        }
    };
}(BitButil));
